### PR TITLE
Fixes #7616: Ads information count

### LINKED ISSUE #7616: FXIOS-1176 ⁃ Ads information counts
**Vision**

As a PM, I want visibility into how often different search activities are taking place within our app to monitor for issues and measure results of improvements to the search workflows

**Acceptance Criteria**

* Counts in Glean for ad clicks and pages with ads
* Data is inline with Fenix for simpler reporting

**Notes**

* Corresponding [Fenix tickets and PRs](https://github.com/mozilla-mobile/fenix/issues/6448)
* [Search definitions](https://docs.telemetry.mozilla.org/datasets/search.html)
* Travis or Megan can help answer questions and ensure tracking is inline with Fenix for simpler reporting

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FXIOS-2625)

### PR DESCRIPTION
**Pull request by nbhasin2**
Whats done:
Basic functionality of where to track the ad clicks have been added. 
-> Search with ads i.e. visit a search provider page that contains ads
-> Open an ad from the search provider page
-> Open an ad in a new tab from the search provider page

Edge case: 
If a user opens a search provider page which displays ads and then in the same tab opens a URL which isn't search provider and taps on an add. In rare scenario if this URL matches our list we count it as an ad click from the provider page.

Remaining:
 ~Find a better place to clear list of ad URLs from search provider~ 
- [x] Investigate better regex for bing and other providers as its adding non-ad urls to the list as well
- [x] Add glean telemetry calls 
- [x] Data review

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift [hunk 1]
```diff
@@ -342,7 +349,15 @@ extension BrowserViewController: WKNavigationDelegate {
             decisionHandler(.cancel)
             return
         }
-
+        
+        if tab == tabManager.selectedTab, navigationAction.navigationType == .linkActivated, tab.adsTelemetryUrlList.count > 0 {
+            let adUrl = url.absoluteString
+            if tab.adsTelemetryUrlList.contains(adUrl) {
+                print("TRACK ADS CLICKED ON PROVIDER PAGE")
```
Comment (nbhasin2) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift:356]: Replace with Glean: TRACK ADS CLICKED ON PROVIDER PAGE

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift [hunk 2]
```diff
@@ -128,6 +128,13 @@ extension BrowserViewController: WKUIDelegate {
                 let elements = contextHelper.elements else { return nil }
             let isPrivate = currentTab.isPrivate
             let addTab = { (rURL: URL, isPrivate: Bool) in
+                if currentTab == self.tabManager.selectedTab, currentTab.adsTelemetryUrlList.count > 0 {
+                    let adUrl = rURL.absoluteString
+                    if currentTab.adsTelemetryUrlList.contains(adUrl) {
+                        print("TRACK ADS CLICKED ON PROVIDER PAGE - OPEN IN NEW TAB")
```
Comment (nbhasin2) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift:134]: Replace with Glean: TRACK ADS CLICKED ON PROVIDER PAGE - OPEN IN NEW TAB

File: Client/Frontend/Browser/BrowserViewController/AdsTelemetryHelper.swift [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        return "adsMessageHandler"
+    }
+
+    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        guard
+            let provider = getProviderForMessage(message: message),
+            let body = message.body as? [String : Any],
+            let urls = body["urls"] as? [String] else { return }
+        let webpageHasAds = provider.containsAds(urls: urls)
+        print("TRACK ADS Found on a provider page ",webpageHasAds)
```
Comment (nbhasin2) [inline Client/Frontend/Browser/BrowserViewController/AdsTelemetryHelper.swift:134]: Replace with Glean: TRACK ADS found on a provider page

**Comment by nbhasin2** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?
    - How often different search activities are taking place within our app 
    - What is the click-through rate for ads on search provider pages

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? Some example responses:
    - These events will help monitor search issues and measure results of improvements to the search workflows

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

   None

4) Can current instrumentation answer these questions?

   No

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

Category 2

<table>
  <tr>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>Track ads click on a search provider page</td>
    <td>Category 2</td>
    <td> https://github.com/mozilla-mobile/firefox-ios/issues/7616 </td>
  </tr>
  <tr>
    <td>Track when user lands on a search provider page which has ads on it</td>
    <td>Category 2</td>
    <td> https://github.com/mozilla-mobile/firefox-ios/issues/7616 </td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
Available in metrics.yaml and https://dictionary.telemetry.mozilla.org/apps/firefox_ios

7) How long will this data be collected?  Choose one of the following:

    Until 2022-02-09 (approximately 6 months) so we can gather enough data

8) What populations will you measure?
All

9) If this data collection is default on, what is the opt-out mechanism for users?
Default on, users can opt out in settings

10) Please provide a general description of how you will analyze this data.
Using Glean / Amplitude

11) Where do you intend to share the results of your analysis?
Only on Glean, Amplitude and with mobile teams.

12) Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection? If so:
None

**Review by boek — APPROVED** [review]
just a rogue print! Data review incoming

**Comment by boek** [comment]
# Data Review Form (to be filled by Data Stewards)

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?
[Yes](https://dictionary.telemetry.mozilla.org/apps/firefox_ios?itemType=metrics&page=1&showExpired=1)

2) Is there a control mechanism that allows the user to turn the data collection on and off?
Yes, in settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?
Has expiry

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?
Type 2

5) Is the data collection request for default-on or default-off?
Default on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?
No

7) Is the data collection covered by the existing Firefox privacy notice? **If unsure: escalate to legal if:**
Yes

8) Does the data collection use a third-party collection tool?
No

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift [hunk 4]
```diff
@@ -342,7 +350,16 @@ extension BrowserViewController: WKNavigationDelegate {
             decisionHandler(.cancel)
             return
         }
-
+        
+        if tab == tabManager.selectedTab, navigationAction.navigationType == .linkActivated, tab.adsTelemetryUrlList.count > 0 {
+            let adUrl = url.absoluteString
+            if tab.adsTelemetryUrlList.contains(adUrl) {
+                if !tab.adsProviderName.isEmpty { AdsTelemetryHelper.trackAdsClickedOnPage(providerName: tab.adsProviderName) }
+                tab.adsTelemetryUrlList.removeAll()
```
Comment (st3fan) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift:358]: @nbhasin2 The `adsTelemetryUrlList` and `adsProviderName` only get reset if the URL that is being navigated to is in the `adsTelemetryUrlList`. Is that correct? Should we not reset those two always after navigating to a new page?

Comment (nbhasin2) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift:358]: @st3fan You are right, in an ideal scenario we should reset `adsTelemetryUrlList` and `adsProviderName` on navigation. 

During my implementation I found that before we land on the final url there are many redirects that happen with search engine ads url before landing on final website and this was the best possible solution at the time to check if an ad url is part of redirects and once found we can then ignore other redirect urls. 

I am open to exploring any idea that you might have to cope with redirects and if Focus does it differently.

Comment (st3fan) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WebViewDelegates.swift:358]: No i think you are right about the redirects

**Pull request merged**